### ISSUE TITLE
What's the performance of a Handle scribe?

### ISSUE THREAD

**Issue by adinapoli** [Soostone/katip#46]
Hey @MichaelXavier !

That's mostly a question rather than anything, but these days I'm taking a stab at implementing a `Scribe` for `syslog` so I dug a bit into the internals of Katip.

Specifically, I have noticed two things (but it wasn't a **deep** delve, so sorry if I'm saying something stupid!):

* There is an `MVar` lock in the `Handle` Scribe implementation, to prevent contention of an `Handle` when lots of concurrency is going on

* The implementation of [logItem](https://github.com/Soostone/katip/blob/master/katip/src/Katip/Core.hs#L695) doesn't use any queue of sort, but it delegates the logging to each scribe in the `LogEnv`. Therefore, I couldn't but wonder if this can hurt the performance in a production environment where you have lots of contention on a handle (say, `stdout`). Wouldn't be better to expand `logItem` in a way that we could keep a buffered queue of the incoming messages and dispatching them along?

I'm sure you guys thought about all of this when fleshing out the API (for example, an unanswered question is what should be the morally thing to do in case you cannot serve the messages fast enough at the Scribe level. Should Katip allow an unbounded buffer in memory? Dump messages?) but I wanted nevertheless spin up a discussion to see what do you guys recommend.

In particular, I would like to avoid a huge performance tax for adding logs to a file 😉 

Cheers!

Alfredo

**Comment by MichaelXavier** [comment]
I'm certain there's low hanging performance fruit for the handle scribe. In fact there's a simple [benchmark](https://github.com/Soostone/katip/blob/master/katip/bench/Main.hs) for the handle scribe to make it easier should somebody want to take a run at improving performance. To give you an idea of the design process, katip was designed specifically with the ES scribe in mind first. The handle scribe came in at the same time because we needed to support file logging but we've not made much of an attempt at optimizing that path because it was fast enough. Scribes are sufficiently generic that that should be pretty amenable to more performant implementations that don't disrupt the end user's API.

Regarding the MVar lock, the discussion in #34 is worth reading: MVar was the quick solution, a bounded queue is probably better. In katip-elasticsearch we use a TBMQueue. I definitely wouldn't want to use any sort of unbounded queue. In my view, at the very least the user should have to opt into an unbounded queue and even then that sounds dubious. It would probably be a good idea for both scribes to use a bounded queue from unagi-chan rather than STM. As far as I know that library is the current performance leader for such things.

tl;dr: if you come up with a more performant logging mechanism for either of the katip scribes and benchmarks to show it, I'm happy. If you also avoid breaking the API while doing so, I'm very happy :)

**Comment by adinapoli** [comment]
Hey Michael!

> To give you an idea of the design process, katip was designed specifically with the ES scribe in mind first.

Makes total sense! After all, at work we use Loggly (same underlying idea, I guess) so it's more than understandable that the Handle scribe is a bit of a "second class citizen" 😉 

> Scribes are sufficiently generic that that should be pretty amenable to more performant implementations that don't disrupt the end user's API.

Aye, although my small pet peeve could be that moving the complexity of handling a queue into the main logging function will (yes) break the user API but it means creating a new `Scribe` will be a matter of simply choosing the format in which you want to log stuff, because the heavy performance lifting will be done for you automatically. But I'm happy even with the idea of having Katip merely off-loading any performance concern to the underlying `Scribe`(s), although that could feel a bit like reinventing the wheel every time.

> MVar was the quick solution, a bounded queue is probably better. In katip-elasticsearch we use a TBMQueue. I definitely wouldn't want to use any sort of unbounded queue. In my view, at the very least the user should have to opt into an unbounded queue and even then that sounds dubious.

Aye, unbounded queues are nasty, and personally I would chose any day losing lose logs than having to be on a call-duty at 3AM because my server crashed due to some memory leak 😛  I guess that for particular hardcore users we could offer an unbounded queue, but still..

> It would probably be a good idea for both scribes to use a bounded queue from unagi-chan rather than STM. As far as I know that library is the current performance leader for such things.

I'm well aware of `unagi-chan` and indeed it has amazing performances, although I have to sheepishly admit I have always stirred away from it as I was put off by these 2 lines in the description:

> The library may be of limited usefulness outside of x86 architectures where the fetch-and-add instruction is not available.

That always gave me the impression using `unagi-chan` for a library was not a good choice as it's locking you down to a particular CPU architecture but I would be very happy to be proven wrong from guys who know more than me on the topic, as I'm begging for an excuse to try it 😉 

> tl;dr: if you come up with a more performant logging mechanism for either of the katip scribes and benchmarks to show it, I'm happy. If you also avoid breaking the API while doing so, I'm very happy :)

Will try to cook up something, but to quote Ed Kmett "I reserve the right to get bored and do something which actually makes money" 😅

**Comment by MichaelXavier** [comment]
I mean even if unagi-chan ends up as slow as TBMQueue when used on other architectures it would still be stricly a win. I think the bar may be fairly low.

Perhaps there's a way to preserve the API while moving buffering into Katip itself so scribes don't have to reinvent the wheel every time: a scribe is `LogItem a => Item a -> IO ()`. That API it seems to me is fully compatible both with pushing to a queue and to receiving from a queue that's managed elsewhere. So what if katip kept a bounded chan of some sort internally. Every log you send it, it tries to write to that channel. Next, when you add a scribe it dups that channel and wires it up to your scribe to fit that api. In that scenario, every call to a log statement would synchronously write to the chan and immediately return, not waiting around for the file write, ES index or whatever else. Removing a scribe I think would have to destroy the reference to the duped chan.

This seems like it would work well provided that:
1. You'd specify a bounding size at the katip level (with maybe a reasonable default).
2. That slow consumers couldn't disrupt fast consumers. Unagi-chan says for dupChan `Writers will be blocked only when the fastest reader falls behind the bounds; slower readers of duplicated OutChan may fall arbitrarily behind.`. It isn't clear to me what "arbitrarily fall behind" means. If your bound is 10, you have 1 consumer  that consumes instantly, one that consumes slowly and you're at 10 on the main chan, does that mean the main chan can accept more as soon as the fast worker is done, but the slow chan gets their own queue of 10, thus the amount of memory you consume is bound * number of dups?
3. It's actually possible to disconnect a chan when its removed and to ensure chans are all flushed on shutdown. I do this in katip-elasticsearch because it provides a "closeable" channel so you can cut off all new entries and wait for it to drain.

I certainly don't want to lay this out at your feet as a user. I'm happy to work any of this but I know my performance techniques in Haskell are not stellar so I'd much prefer someone who really knows their stuff tell me about the land mines before I go for a stroll in the field.

**Comment by adinapoli** [comment]
> I mean even if unagi-chan ends up as slow as TBMQueue when used on other architectures it would still be stricly a win. I think the bar may be fairly low.

Ok, I'm sold 😅 

> That API it seems to me is fully compatible both with pushing to a queue and to receiving from a queue that's managed elsewhere. 

Yes.

> So what if katip kept a bounded chan of some sort internally. Every log you send it, it tries to write to that channel. [..]

At work we have been working together with Well Typed on our backend server, and they came up (summoning the tremendously productive @adamgundry) with a logging system which fairly resemble Katip and I think it's working well enough maybe we can shamelessly steal ideas? 😉 

Last time I checked, this logging system was based on mainly 2 concepts: a `LogContext` (the equivalent of our `LogEnv` plus some other stuff for bookeeping) and the concept of a `Driver` (`Scribe` anyone?). The idea is that _each_ `Driver` would get a separate (bounded) queue where messages will be written by the main code loop. One very nice thing a `Driver` has is that it will
"disable itself" if the queue fills up and therefore is slowing the entire logging system down. I would envision something like this to work possibly better than trying to get right the entire channel duping, which can be a bit insidious at time, but maybe I am being a bit too pessimistic here! Maybe Adam can chime and help us out as he solved the problem in the past already?

> I certainly don't want to lay this out at your feet as a user. I'm happy to work any of this but I know my performance techniques in Haskell are not stellar so I'd much prefer someone who really knows their stuff tell me about the land mines before I go for a stroll in the field.

I feel I'm on the same boat as you are, I certainly can survive a battle but not the war in the performance minefield 😁  Said that, I'm more than happy to review the code if you come up with a PR or take the plunge myself casually during hacking hours provided I can scrap my work at any time if you arrive first. TL;DR Don't feel blocked by the fact I could potentially take a stab at it!

As this thread is potentially shaping up to be something "bigger" than addressing only the Handle Scribe performance, I think I will take a stab at the simplest thing we originally suggested (using a queue for the Handle Scribe and benchmark it) and worst case scenario submit a quick cheeky patch so I can still benefit from the speed boost immediately while we wait for design decision of this ticket to settle down 😉

**Comment by adamgundry** [comment]
@adinapoli your description of what we do sounds about right to me. The key ideas are these:
 * Each driver/scribe has its own bounded queue of configurable length, with a thread that reads from this queue and does something with the logs.
 * When an application thread generates log messages, it writes them into all the queues without blocking. If a queue is full, that log driver gets shut down (and a warning is logged!).

One could of course choose to discard log messages sent to slow drivers but give them a chance to catch up and start logging again.

**Comment by MichaelXavier** [comment]
@adamgundry Interesting. So when you say a driver gets shut down, is it permanent or does it get reinstated at some later point if it finishes processing its backlog?

**Comment by adamgundry** [comment]
In our application, it's permanent (the driver is flagged as broken and we will no longer write messages to its queue). Though as I say, one could also merely temporarily stop sending log items while the queue is full, but start again when there is space. Essentially this amounts to making drivers lossy if they can't keep up a high enough throughput.

**Comment by MichaelXavier** [comment]
Thanks for the clarification, @adamgundry. I'm tempted to go the route of just dropping messages if the queue can't keep up. I've also created #49 for future discussion of this queueing architecture. We can have the solution to this issue be @adinapoli 's #48 PR. I discuss some possible breaking changes around finalizers in #49 so it may even be preferable to merge #48, let you guys play around with it unreleased on hackage and then release when #49 is solved to cut down on intermediate API breakages.

**Comment by adinapoli** [comment]
That sounds great guys! Something I would add very casually is that both `TBQueue` and Unagi's have something like `tryWriteXXX` which allows you to write something in the queue in a "best effort" manner; if the queue is full, the message will be dropped, but if the queue starts to drain the Scribe will "reactivate" itself automatically, which is a nice thing to have, methinks.

**Comment by MichaelXavier** [comment]
Yeah that's exactly what I had in mind. Seems like it gets you pretty far. I think it may even better than a more complicated backoff strategy because I  think its pretty cheap to to trryWrite and you're guaranteed to resume normal operation as soon as possible (if possible).

**Comment by MichaelXavier** [comment]
This was released some time ago.
